src/main/java/build/buildfarm/instance/shard/ShardInstance.java

```diff
@@ -1904,7 +1911,7 @@ private void deliverCachedActionResult(
       Operation operation,
       RequestMetadata requestMetadata)
       throws Exception {
-    recentCacheServedExecutions.put(requestMetadata, true);
+    recentCacheServedExecutions.put(requestMetadata, CompletableFuture.completedFuture(true));
```

Comment:
unnecessary once you swap recentCacheServedExecutions to be synchronous.

Reply:
fixed